src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,406 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::collections::Bound::{self, Excluded, Included, Unbounded};
+use std::default::Default;
+use std::fmt::{self, Debug, Formatter};
+use std::ops::RangeBounds;
+use std::sync::{Arc, RwLock};
+
+use kvproto::kvrpcpb::Context;
+
+use raftstore::store::engine::IterOption;
+use storage::engine::{
+    Callback as EngineCallback, CbContext, Cursor, Engine, Error as EngineError, Iterator, Modify,
+    Result as EngineResult, ScanMode, Snapshot,
+};
+use storage::{CfName, Key, Value, CF_DEFAULT, CF_LOCK, CF_WRITE};
+
+type RwLockTree = RwLock<BTreeMap<Key, Value>>;
+
+#[derive(Clone)]
+pub struct BTreeEngine {
```

Comment:
Maybe add a comment to emphasize that this engine is in memory and only used in tests or layered benchmarks.